lib/solidus_segment/serializers/product_serializer.rb

```diff
@@ -3,23 +3,29 @@
 module SolidusSegment
   module Serializers
     class ProductSerializer
-      def initialize(line_item)
+      # The line item is the product specialization during
+      # the checkout process in Solidus so we should use it if present.
+      def initialize(variant:, line_item: nil)
+        @variant = variant
         @line_item = line_item
       end
 
       def to_h
         {
-          id: line_item.variant.id,
-          sku: line_item.variant.sku,
-          name: line_item.product.name,
-          price: line_item.price,
-          quantity: line_item.quantity,
+          id: variant.id,
+          sku: variant.sku,
+          name: variant.product.name,
+          variant: variant.options_text,
+          price: line_item&.price || variant.price,
+          quantity: line_item&.quantity || 1,
+          currency: line_item&.currency || variant.default_price.currency,
+          value: line_item&.total || variant.price
```

Comment:
Instead of inlining these many conditionals would it make sense to replace them with polymorphism and use two different serializers? `ProductSerializer` could internally create and use a `VariantSerializer` or a `LineItemSerializer` depending on its needs.

Reply:
In my opinion, the code could be clearer using three different serializers as you wrote above but I'm a little skeptical about this approach since we expose two more APIs that shouldn't be used.

Overall, for me its the same! I'll refactor the PR ASAP.